sycl/test-e2e/VirtualMem/virtual_mem_operations.cpp

```diff
@@ -0,0 +1,122 @@
+    syclext::free_virtual_mem(VirtualMemoryPtr, AlignedByteSize, Context);
+  }
+
+  {
+    // Check should be performed that methods get_context(), get_device() and
+    // size() return correct values (i.e. ones which were passed to physical_mem
+    // constructor).
+    size_t BytesRequired = NumberOfElementsPerIteration[2] * sizeof(int);
+    size_t UsedGranularity = GetLCMGranularity(Device, Context);
+    size_t AlignedByteSize = GetAlignedByteSize(BytesRequired, UsedGranularity);
+
+    syclext::physical_mem PhysicalMem{Device, Context, AlignedByteSize};
+
+    PhysicalMem.get_context();
+    PhysicalMem.get_device();
+    PhysicalMem.size();
+
+    assert(PhysicalMem.get_device() == Device &&
+           "device passed to physical_mem must be the same as returned from "
+           "get_device()");
+
+    assert(PhysicalMem.get_context() == Context &&
+           "context passed to physical_mem must be the same as returned from "
+           "get_context()");
+
+    assert(PhysicalMem.size() == AlignedByteSize &&
+           "size in bytes passed to physical_mem must be the same as returned "
+           "from size()");
+  }
+
+  {
+    // Check to see if value returned from a valid call to map() is the same as
+    // reinterpret_cast<void *>(ptr).
+    size_t BytesRequired = NumberOfElementsPerIteration[2] * sizeof(int);
+    size_t UsedGranularity = GetLCMGranularity(Device, Context);
+    size_t AlignedByteSize = GetAlignedByteSize(BytesRequired, UsedGranularity);
+
+    uintptr_t VirtualMemoryPtr =
+        syclext::reserve_virtual_mem(0, AlignedByteSize, Context);
+
+    syclext::physical_mem PhysicalMem{Device, Context, AlignedByteSize};
+
+    void *MappedPtr = PhysicalMem.map(VirtualMemoryPtr, AlignedByteSize,
+                                      syclext::address_access_mode::read_write);
+
+    assert(MappedPtr == reinterpret_cast<void *>(VirtualMemoryPtr) &&
+           "value returned from a valid call to map() must be equal "
+           "reinterpret_cast<void *>(ptr)");
+
+    syclext::unmap(MappedPtr, AlignedByteSize, Context);
+    syclext::free_virtual_mem(VirtualMemoryPtr, AlignedByteSize, Context);
+  }
+
+  // Check to see if can change access mode of a virtual memory range and
+  // immediately see it changed.
+  for (const size_t RequiredNumElements : NumberOfElementsPerIteration) {
+    size_t BytesRequired = RequiredNumElements * sizeof(int);
+    size_t UsedGranularity = GetLCMGranularity(Device, Context);
+    size_t AlignedByteSize = GetAlignedByteSize(BytesRequired, UsedGranularity);
+    uintptr_t VirtualMemoryPtr =
+        syclext::reserve_virtual_mem(0, AlignedByteSize, Context);
+    syclext::physical_mem PhysicalMem{Device, Context, AlignedByteSize};
+    void *MappedPtr = PhysicalMem.map(VirtualMemoryPtr, AlignedByteSize,
+                                      syclext::address_access_mode::read_write);
+
+    syclext::set_access_mode(MappedPtr, AlignedByteSize,
```

Comment:
Can we also check that before this change the access mode is `read_write` as set in `map`? Just to make sure that it wasn't `read` all the time.

Reply:
Sure, will be done.